"unknown option `--root'" error with RN@0.57-rc during packager startup

## Actual Behavior
1. Opening command palette, selecting start pakager fails to run packager.
2. It also shows a "unknown option --root" in the terminal

## Expected Behavior
1. Packager should be launched instantly

## Software versions
- React-Native VS Code extension version: 0.6.16
- VSCode version: 1.27.1 (User Setup)
- OS platform and version: Windows 10 x64 v 16299.611
- NodeJS version: 8.11.4
- React Native version: 0.57.0-rc.4
- Expo SDK version (include if relevant):

## Outputs (Include if relevant)
- Output of the Debug Console (View -> Toggle Debug Console):
```
Nothing
```
- Output of the React-Native output channel (View -> Toggle Output -> Select React-Native in ListBox):
```
[Info] Starting Packager


  error: unknown option `--root'

[Error] Failed to start the React Native packager

Stack: Error: Failed to start the React Native packager: Could not start the packager.
    at Function.getWrappedError (C:\Users\reza\.vscode\extensions\vsmobile.vscode-react-native-0.6.16\src\common\error\internalError.js:35:16)
    at Function.wrapError (C:\Users\reza\.vscode\extensions\vsmobile.vscode-react-native-0.6.16\src\common\error\errorHelper.js:18:44)
    at resultOfCode.done.reason (C:\Users\reza\.vscode\extensions\vsmobile.vscode-react-native-0.6.16\src\common\entryPointHandler.js:43:72)
    at _rejected (C:\Users\reza\.vscode\extensions\vsmobile.vscode-react-native-0.6.16\node_modules\q\q.js:844:24)
    at C:\Users\reza\.vscode\extensions\vsmobile.vscode-react-native-0.6.16\node_modules\q\q.js:870:30
    at Promise.when (C:\Users\reza\.vscode\extensions\vsmobile.vscode-react-native-0.6.16\node_modules\q\q.js:1122:31)
    at Promise.promise.promiseDispatch (C:\Users\reza\.vscode\extensions\vsmobile.vscode-react-native-0.6.16\node_modules\q\q.js:788:41)
    at C:\Users\reza\.vscode\extensions\vsmobile.vscode-react-native-0.6.16\node_modules\q\q.js:604:44
    at runSingle (C:\Users\reza\.vscode\extensions\vsmobile.vscode-react-native-0.6.16\node_modules\q\q.js:137:13)
    at flush (C:\Users\reza\.vscode\extensions\vsmobile.vscode-react-native-0.6.16\node_modules\q\q.js:125:13)
    at _combinedTickCallback (internal/process/next_tick.js:131:7)
    at process._tickCallback (internal/process/next_tick.js:180:9)
```
- Output of the Developer Tools console (Help -> Toggle Developer Tools -> Select Console tab):
```
[Extension Host] (node:13544) [DEP0022] DeprecationWarning: os.tmpDir() is deprecated. Use os.tmpdir() instead.
t.log @ console.ts:136
t._logExtensionHostMessage @ extensionHost.ts:412
(anonymous) @ extensionHost.ts:224
emitTwo @ events.js:126
emit @ events.js:214
emit @ internal/child_process.js:772
_combinedTickCallback @ internal/process/next_tick.js:141
_tickCallback @ internal/process/next_tick.js:180
notificationsAlerts.ts:39 Error: Could not start the packager.
    at Q.then.then.conditionResult (C:\Users\reza\.vscode\extensions\vsmobile.vscode-react-native-0.6.16\src\common\node\promise.js:52:23)
    at _fulfilled (C:\Users\reza\.vscode\extensions\vsmobile.vscode-react-native-0.6.16\node_modules\q\q.js:834:54)
    at self.promiseDispatch.done (C:\Users\reza\.vscode\extensions\vsmobile.vscode-react-native-0.6.16\node_modules\q\q.js:863:30)
    at Promise.promise.promiseDispatch (C:\Users\reza\.vscode\extensions\vsmobile.vscode-react-native-0.6.16\node_modules\q\q.js:796:13)
    at C:\Users\reza\.vscode\extensions\vsmobile.vscode-react-native-0.6.16\node_modules\q\q.js:556:49
    at runSingle (C:\Users\reza\.vscode\extensions\vsmobile.vscode-react-native-0.6.16\node_modules\q\q.js:137:13)
    at flush (C:\Users\reza\.vscode\extensions\vsmobile.vscode-react-native-0.6.16\node_modules\q\q.js:125:13)
    at _combinedTickCallback (internal/process/next_tick.js:131:7)
    at process._tickCallback (internal/process/next_tick.js:180:9)
```
RN 0.56 had some problems in windows, it has been fixed in 0.57-rc, but I cant run metro bundler from inside of vscode now. If I launch metro manually, debugger and other features will work as expected.

--- a/src/common/packager.ts
+++ b/src/common/packager.ts
@@ -84,7 +84,7 @@ export class Packager {
                     .then((version) => {
                         return helper.isExpoApp(false)
                         .then((isExpo) => {
-                            if (!isExpo) {
+                            if (isExpo) {
                                 // Arguments below using for Expo apps
 
                                 if (!semver.gte(version, "0.57.0")) {

--- a/src/common/packager.ts
+++ b/src/common/packager.ts
@@ -35,10 +35,12 @@ export class Packager {
     private static REACT_NATIVE_PACKAGE_NAME = "react-native";
     private static OPN_PACKAGE_MAIN_FILENAME = "index.js";
     private static fs: FileSystem = new Node.FileSystem();
+    private expoHelper: ExponentHelper;
 
     constructor(private workspacePath: string, private projectPath: string, private packagerPort?: number, packagerStatusIndicator?: PackagerStatusIndicator) {
         this.packagerStatus = PackagerStatus.PACKAGER_STOPPED;
         this.packagerStatusIndicator = packagerStatusIndicator || new PackagerStatusIndicator();
+        this.expoHelper = new ExponentHelper(this.workspacePath, this.projectPath);
     }
 
     public get port(): number {
@@ -60,83 +62,84 @@ export class Packager {
         return this.packagerStatus;
     }
 
+    public getPackagerArgs(rnVersion: string, resetCache: boolean = false): Q.Promise<string[]> {
+        let args: string[] = ["--port", this.port.toString()];
+
+        if (resetCache) {
+            args = args.concat("--resetCache");
+        }
+
+        return this.expoHelper.isExpoApp(false)
+        .then((isExpo) => {
+            if (!isExpo) {
+                return args;
+            }
+
+            // Arguments below using for Expo apps
+            if (!semver.gte(rnVersion, "0.57.0")) {
+                args.push("--root", path.relative(this.projectPath, path.resolve(this.workspacePath, ".vscode")));
+            }
+
+            return this.expoHelper.getExpPackagerOptions()
+            .then((options: ExpConfigPackager) => {
+                Object.keys(options).forEach(key => {
+                    args = args.concat([`--${key}`, options[key]]);
+                });
+
+                return args;
+            })
+            .catch(() => {
+                this.logger.warning("Couldn't read packager's options from exp.json, continue...");
+
+                return args;
+            });
+        });
+    }
+
     public start(resetCache: boolean = false): Q.Promise<void> {
         this.packagerStatusIndicator.updatePackagerStatus(PackagerStatus.PACKAGER_STARTING);
         let executedStartPackagerCmd = false;
+        let rnVersion: string;
 
         return this.isRunning()
         .then((running) => {
-            if (!running) {
-                executedStartPackagerCmd = true;
-
-                return this.monkeyPatchOpnForRNPackager()
-                .then(() => {
-                    let args: string[] = ["--port", this.port.toString()];
-                    if (resetCache) {
-                        args = args.concat("--resetCache");
-                    }
-
-                    let helper = new ExponentHelper(this.workspacePath, this.projectPath);
-
-                    const projectRoot = SettingsHelper.getReactNativeProjectRoot(this.workspacePath);
-
-                    return ReactNativeProjectHelper.getReactNativeVersion(projectRoot)
-                    .then((version) => {
-                        return helper.isExpoApp(false)
-                        .then((isExpo) => {
-                            if (isExpo) {
-                                // Arguments below using for Expo apps
-
-                                if (!semver.gte(version, "0.57.0")) {
-                                    args.push("--root", path.relative(this.projectPath, path.resolve(this.workspacePath, ".vscode")));
-                                }
-
-                                return helper.getExpPackagerOptions()
-                                .then((options: ExpConfigPackager) => {
-                                    Object.keys(options).forEach(key => {
-                                        args = args.concat([`--${key}`, options[key]]);
-                                    });
-
-                                    return args;
-                                })
-                                .catch(() => {
-                                    this.logger.warning("Couldn't read packager's options from exp.json, continue...");
-
-                                    return args;
-                                });
-                            } else {
-                                return args;
-                            }
-                        })
-                        .then((args) => {
-                            //  There is a bug with launching VSCode editor for file from stack frame in 0.38, 0.39, 0.40 versions:
-                            //  https://github.com/facebook/react-native/commit/f49093f39710173620fead6230d62cc670570210
-                            //  This bug will be fixed in 0.41
-                            const failedRNVersions: string[] = ["0.38.0", "0.39.0", "0.40.0"];
-
-                            let reactEnv = Object.assign({}, process.env, {
-                                REACT_DEBUGGER: "echo A debugger is not needed: ",
-                                REACT_EDITOR: failedRNVersions.indexOf(version) < 0 ? "code" : this.openFileAtLocationCommand(),
-                            });
-
-                            this.logger.info("Starting Packager");
-                            // The packager will continue running while we debug the application, so we can"t
-                            // wait for this command to finish
-
-                            let spawnOptions = { env: reactEnv };
-
-                            const packagerSpawnResult = new CommandExecutor(this.projectPath, this.logger).spawnReactPackager(args, spawnOptions);
-                            this.packagerProcess = packagerSpawnResult.spawnedProcess;
-                            packagerSpawnResult.outcome.done(() => { }, () => { }); // Q prints a warning if we don't call .done(). We ignore all outcome errors
-
-                            return packagerSpawnResult.startup;
-                        });
-                    });
-
-                });
-            } else {
+            if (running) {
                 return void 0;
             }
+
+            executedStartPackagerCmd = true;
+
+            return this.monkeyPatchOpnForRNPackager()
+            .then(() => {
+                return ReactNativeProjectHelper.getReactNativeVersion(this.projectPath);
+            })
+            .then((version) => {
+                rnVersion = version;
+                return this.getPackagerArgs(rnVersion, resetCache);
+            })
+            .then((args) => {
+                //  There is a bug with launching VSCode editor for file from stack frame in 0.38, 0.39, 0.40 versions:
+                //  https://github.com/facebook/react-native/commit/f49093f39710173620fead6230d62cc670570210
+                //  This bug will be fixed in 0.41
+                const failedRNVersions: string[] = ["0.38.0", "0.39.0", "0.40.0"];
+
+                let reactEnv = Object.assign({}, process.env, {
+                    REACT_DEBUGGER: "echo A debugger is not needed: ",
+                    REACT_EDITOR: failedRNVersions.indexOf(rnVersion) < 0 ? "code" : this.openFileAtLocationCommand(),
+                });
+
+                this.logger.info("Starting Packager");
+                // The packager will continue running while we debug the application, so we can"t
+                // wait for this command to finish
+
+                let spawnOptions = { env: reactEnv };
+
+                const packagerSpawnResult = new CommandExecutor(this.projectPath, this.logger).spawnReactPackager(args, spawnOptions);
+                this.packagerProcess = packagerSpawnResult.spawnedProcess;
+                packagerSpawnResult.outcome.done(() => { }, () => { }); // Q prints a warning if we don't call .done(). We ignore all outcome errors
+
+                return packagerSpawnResult.startup;
+            });
         })
         .then(() => {
             return this.awaitStart();

--- a/src/common/packager.ts
+++ b/src/common/packager.ts
@@ -18,6 +18,7 @@ import {SettingsHelper} from "../extension/settingsHelper";
 import * as Q from "q";
 import * as path from "path";
 import * as XDL from "../extension/exponent/xdlInterface";
+import * as semver from "semver";
 import { FileSystem } from "./node/fileSystem";
 
 export class Packager {
@@ -62,85 +63,96 @@ export class Packager {
     public start(resetCache: boolean = false): Q.Promise<void> {
         this.packagerStatusIndicator.updatePackagerStatus(PackagerStatus.PACKAGER_STARTING);
         let executedStartPackagerCmd = false;
+
         return this.isRunning()
-            .then(running => {
-                if (!running) {
-                    executedStartPackagerCmd = true;
-                    return this.monkeyPatchOpnForRNPackager()
-                        .then(() => {
-                            let args: string[] = ["--port", this.port.toString()];
-                            if (resetCache) {
-                                args = args.concat("--resetCache");
-                            }
+        .then((running) => {
+            if (!running) {
+                executedStartPackagerCmd = true;
 
-                            let helper = new ExponentHelper(this.workspacePath, this.projectPath);
+                return this.monkeyPatchOpnForRNPackager()
+                .then(() => {
+                    let args: string[] = ["--port", this.port.toString()];
+                    if (resetCache) {
+                        args = args.concat("--resetCache");
+                    }
 
-                            return helper.isExpoApp(false)
-                            .then((isExpo) => {
-                                if (isExpo) {
-                                    // Arguments below using for Expo apps
+                    let helper = new ExponentHelper(this.workspacePath, this.projectPath);
+
+                    const projectRoot = SettingsHelper.getReactNativeProjectRoot(this.workspacePath);
+
+                    return ReactNativeProjectHelper.getReactNativeVersion(projectRoot)
+                    .then((version) => {
+                        return helper.isExpoApp(false)
+                        .then((isExpo) => {
+                            if (!isExpo) {
+                                // Arguments below using for Expo apps
+
+                                if (!semver.gte(version, "0.57.0")) {
                                     args.push("--root", path.relative(this.projectPath, path.resolve(this.workspacePath, ".vscode")));
-                                    return helper.getExpPackagerOptions()
-                                    .then((options: ExpConfigPackager) => {
-                                        Object.keys(options).forEach(key => {
-                                            args = args.concat([`--${key}`, options[key]]);
-                                        });
-
-                                        return args;
-                                    })
-                                    .catch(() => {
-                                        this.logger.warning("Couldn't read packager's options from exp.json, continue...");
-                                        return args;
-                                    });
-                                } else {
-                                    return args;
                                 }
-                            });
+
+                                return helper.getExpPackagerOptions()
+                                .then((options: ExpConfigPackager) => {
+                                    Object.keys(options).forEach(key => {
+                                        args = args.concat([`--${key}`, options[key]]);
+                                    });
+
+                                    return args;
+                                })
+                                .catch(() => {
+                                    this.logger.warning("Couldn't read packager's options from exp.json, continue...");
+
+                                    return args;
+                                });
+                            } else {
+                                return args;
+                            }
                         })
                         .then((args) => {
-                            const projectRoot = SettingsHelper.getReactNativeProjectRoot(this.workspacePath);
-                            ReactNativeProjectHelper.getReactNativeVersion(projectRoot).then(version => {
+                            //  There is a bug with launching VSCode editor for file from stack frame in 0.38, 0.39, 0.40 versions:
+                            //  https://github.com/facebook/react-native/commit/f49093f39710173620fead6230d62cc670570210
+                            //  This bug will be fixed in 0.41
+                            const failedRNVersions: string[] = ["0.38.0", "0.39.0", "0.40.0"];
 
-                                //  There is a bug with launching VSCode editor for file from stack frame in 0.38, 0.39, 0.40 versions:
-                                //  https://github.com/facebook/react-native/commit/f49093f39710173620fead6230d62cc670570210
-                                //  This bug will be fixed in 0.41
-                                const failedRNVersions: string[] = ["0.38.0", "0.39.0", "0.40.0"];
-
-                                let reactEnv = Object.assign({}, process.env, {
-                                    REACT_DEBUGGER: "echo A debugger is not needed: ",
-                                    REACT_EDITOR: failedRNVersions.indexOf(version) < 0 ? "code" : this.openFileAtLocationCommand(),
-                                });
-
-                                this.logger.info("Starting Packager");
-                                // The packager will continue running while we debug the application, so we can"t
-                                // wait for this command to finish
-
-                                let spawnOptions = { env: reactEnv };
-
-                                const packagerSpawnResult = new CommandExecutor(this.projectPath, this.logger).spawnReactPackager(args, spawnOptions);
-                                this.packagerProcess = packagerSpawnResult.spawnedProcess;
-                                packagerSpawnResult.outcome.done(() => { }, () => { }); /* Q prints a warning if we don't call .done().
-                                                                                        We ignore all outcome errors */
-                                return packagerSpawnResult.startup;
+                            let reactEnv = Object.assign({}, process.env, {
+                                REACT_DEBUGGER: "echo A debugger is not needed: ",
+                                REACT_EDITOR: failedRNVersions.indexOf(version) < 0 ? "code" : this.openFileAtLocationCommand(),
                             });
+
+                            this.logger.info("Starting Packager");
+                            // The packager will continue running while we debug the application, so we can"t
+                            // wait for this command to finish
+
+                            let spawnOptions = { env: reactEnv };
+
+                            const packagerSpawnResult = new CommandExecutor(this.projectPath, this.logger).spawnReactPackager(args, spawnOptions);
+                            this.packagerProcess = packagerSpawnResult.spawnedProcess;
+                            packagerSpawnResult.outcome.done(() => { }, () => { }); // Q prints a warning if we don't call .done(). We ignore all outcome errors
+
+                            return packagerSpawnResult.startup;
                         });
-                }
+                    });
+
+                });
+            } else {
                 return void 0;
-            })
-            .then(() =>
-                this.awaitStart())
-            .then(() => {
-                this.packagerStatusIndicator.updatePackagerStatus(PackagerStatus.PACKAGER_STARTED);
-                if (executedStartPackagerCmd) {
-                    this.logger.info("Packager started.");
-                    this.packagerStatus = PackagerStatus.PACKAGER_STARTED;
-                } else {
-                    this.logger.info("Packager is already running.");
-                    if (!this.packagerProcess) {
-                        this.logger.warning(ErrorHelper.getWarning("React Native Packager running outside of VS Code. If you want to debug please use the 'Attach to packager' option"));
-                    }
+            }
+        })
+        .then(() => {
+            return this.awaitStart();
+        })
+        .then(() => {
+            this.packagerStatusIndicator.updatePackagerStatus(PackagerStatus.PACKAGER_STARTED);
+            if (executedStartPackagerCmd) {
+                this.logger.info("Packager started.");
+                this.packagerStatus = PackagerStatus.PACKAGER_STARTED;
+            } else {
+                this.logger.info("Packager is already running.");
+                if (!this.packagerProcess) {
+                    this.logger.warning(ErrorHelper.getWarning("React Native Packager running outside of VS Code. If you want to debug please use the 'Attach to packager' option"));
                 }
-            });
+            }
+        });
     }
 
     public stop(silent: boolean = false): Q.Promise<void> {

--- a/test/extension/packager.test.ts
+++ b/test/extension/packager.test.ts
@@ -3,6 +3,7 @@
 
 import { Packager } from "../../src/common/packager";
 import { Request } from "../../src/common/node/request";
+import { ExponentHelper } from "../../src/extension/exponent/exponentHelper";
 
 import * as assert from "assert";
 import * as sinon from "sinon";
@@ -10,13 +11,19 @@ import * as sinon from "sinon";
 suite("packager", function () {
     suite("extensionContext", function () {
         let requestStub: Sinon.SinonStub;
+        let isExpoAppStub: Sinon.SinonStub;
+        let getExpPackagerOptionsStub: Sinon.SinonStub;
 
         setup(() => {
             requestStub = sinon.stub(Request, "request");
+            isExpoAppStub = sinon.stub(ExponentHelper.prototype, "isExpoApp");
+            getExpPackagerOptionsStub = sinon.stub(ExponentHelper.prototype, "getExpPackagerOptions");
         });
 
         teardown(() => {
             requestStub.restore();
+            isExpoAppStub.restore();
+            getExpPackagerOptionsStub.restore();
         });
 
         test("isRunning should check correct status URL", function (done) {
@@ -55,6 +62,70 @@ suite("packager", function () {
                 .then(done, () => {
                     assert.fail(null, null, "packager was not expected to be running");
                     done();
+                });
+        });
+
+        test("getPackagerArgs should return correct value (react-native@0.56.0)", function (done) {
+            isExpoAppStub.returns(Promise.resolve(false));
+            const rnVersion = "0.56.0";
+            const expected = ["--port", "10001"];
+            new Packager("/workspace", "/workspace", 10001)
+                .getPackagerArgs(rnVersion)
+                .then((args) => {
+                    assert.deepEqual(args, expected);
+                    done();
+                })
+                .catch((err) => {
+                    done(err);
+                });
+        });
+
+        test("getPackagerArgs should return correct value (react-native@0.57.0)", function (done) {
+            isExpoAppStub.returns(Promise.resolve(false));
+            const rnVersion = "0.57.0";
+            const expected = ["--port", "10001", "--resetCache"];
+            new Packager("/workspace", "/workspace", 10001)
+                .getPackagerArgs(rnVersion, true)
+                .then((args) => {
+                    assert.deepEqual(args, expected);
+                    done();
+                })
+                .catch((err) => {
+                    done(err);
+                });
+        });
+
+        test("getPackagerArgs should return correct value for expo app (react-native@0.56.0)", function (done) {
+            isExpoAppStub.returns(Promise.resolve(true));
+            getExpPackagerOptionsStub.returns(Promise.resolve({}));
+            const rnVersion = "0.56.0";
+            const expected = ["--port", "10001", "--resetCache", "--root", ".vscode"];
+            new Packager("/workspace", "/workspace", 10001)
+                .getPackagerArgs(rnVersion, true)
+                .then((args) => {
+                    assert.deepEqual(args, expected);
+                    done();
+                })
+                .catch((err) => {
+                    done(err);
+                });
+        });
+
+        test("getPackagerArgs should return correct value for expo app (react-native@0.57.0)", function (done) {
+            isExpoAppStub.returns(Promise.resolve(true));
+            getExpPackagerOptionsStub.returns(Promise.resolve({
+                assetExts: ["txt", "md"],
+            }));
+            const rnVersion = "0.57.0";
+            const expected = ["--port", "10001", "--assetExts", [ "txt", "md"]];
+            new Packager("/workspace", "/workspace", 10001)
+                .getPackagerArgs(rnVersion)
+                .then((args) => {
+                    assert.deepEqual(args, expected);
+                    done();
+                })
+                .catch((err) => {
+                    done(err);
                 });
         });
     });

